### DISCUSSION TITLE
HTTP(S) Websocket  connection possible?

### DISCUSSION THREAD

**Discussion by jonnycastaway** [arendst/Tasmota#21466] [General]
Hi @all,

actually i poll the API to get my SmartMeter (ESP32C3) values and i **don't** want MQTT!
As this is a huge load for the ESP32C3 i'll change to use Websocket.
I use this for other Systems (OpenDTU, Zigbee2MQTT)  and it worked perfectly.
So is it possible to establish a http(s) Websocket connection to Tasmota so i get the Values nearly instantly as they change?

I found this old issue but the issue was closed and conversation died down.
[WebSocket #3368](https://github.com/arendst/Tasmota/issues/3368)

**Comment by sfromis** [comment]
First, using MQTT is **by far** not a "huge load" for a reasonably capable MCU, as this is a pretty lightweight protocol, which works great for reporting values as things happens.

And no, there has not been much interest in WebSockets, as evidenced by you only finding a very old thread without any actual discussion.

When you have ESP32-C3, you can use the Berry language with already supported protocols, or even create new ones. A thing which exists is an implementation of EventSource streaming events to a browser.

If someone wants to work on implementing the WebSockets protocol, using Berry would be a good basis for that.

**Reply by jonnycastaway** [reply]
Hi and thanks for your answer.
I've never sawn that MQTT is a huge load. I don't want MQTT.
I saw that polling the API every second or so is a huge load and i didn't get an reliable answer.
Websocket is a common protocol for nearly Real Time Updates to a browser.
But i'll take a look at EventSource Streaming because it looks like the right answer. It's enough that the Server streams unidirectional. Is there any documentation how to use that with tasmota?

**Reply by sfromis** [reply]
EventSource is not an externalized/documented Tasmota feature, but is being used for the GPIO Viewer feature. Basically just http with multiple writes. The main implementation is in the browser for handling the data.
https://github.com/arendst/Tasmota/tree/development/tasmota/berry/gpio_viewer

**Reply by joba-1** [reply]
Hi, if you don't want MQTT that is your choice, but for anyone else reading this: I read my energy meter roughly every second with an ESP and send it to my MQTT broker which delivers it to several interested parties. Basically no load noticable on the ESP or broker and no dropped packages in years in an environment with lots of other MQTT clients.

**Reply by jonnycastaway** [reply]
> Hi, if you don't want MQTT that is your choice, but for anyone else reading this: I read my energy meter roughly every second with an ESP and send it to my MQTT broker which delivers it to several interested parties. Basically no load noticable on the ESP or broker and no dropped packages in years in an environment with lots of other MQTT clients.

But you have to operate an MQTT Broker. So you have to use some kind of PC or Raspberry Pi etc. and not with an SD-Card because they are broken fast. So SSD is required. 
But what i want is an ESP32 that collect the data from Smartmeter and OpenDTU to vizualize the data on a Display and regulate the inverters.
And that's a task where i would not rely on a MQTT Broker, an extra PC/Pi, an ioBroker etc., an Node-Red or javascript Adapter and so on. This has to be done straight away (imho) and it's working with the API. But Websocket etc. would be better.
So i take a look at Event Source Streaming.
And yes, i used MQTT also some years ago and it's ok as long as it runs. But it's not necessary with all that bloated stuff around it.

Have all a nice weekend

**Reply by sfromis** [reply]
Since OpenDTU seems to support a web API with GET/POST commands, it would appear to be pretty simple to use that from some Berry code when you want to send data. Just make sure that your OpenDTU device can respond reasonably quickly, as the `webclient` module is blocking code until the server responds.

**Comment by Jason2866** [comment]
Not wanting mqtt does mean Tasmota is not the 1st choice. The "m" in Tasmota stands for mqtt.
Tasmota is designed and build for use with mqtt!

**Reply by jonnycastaway** [reply]
Interesting, that's new to me.
Anyway, i can't understand why this feature (webbsocket) is not implemented because it's also usable to update the webui also.
But as i say'd earlier, i'm looking at EventSource.

**Reply by s-hadinger** [reply]
If you want to write a websocket server for esp-idf, please proceed. I haven't found any

**Reply by sfromis** [reply]
Googling for "esp32 websocket server" does have hits which looks relevant, if *someone* wants to take a stab at making something work....

**Reply by s-hadinger** [reply]
The only implementation I have found is [ESPAsyncWebServer](https://github.com/me-no-dev/ESPAsyncWebServer) which is not an option. It's not maintained anymore and is not stable enough

**Reply by Jason2866** [reply]
There is an well maintained fork of [ESPAsyncWebServer](https://github.com/mathieucarbou/ESPAsyncWebServer) but still not an option for Tasmota

**Reply by joba-1** [reply]
still good for me you mentioned this fork, I wasn’t aware of it.
I implemented a real time updating scrollbar that I wanted to use for dimming and similar stuff in my firmwares.
It needs a fast multi request async webserver to react smoothly enough and the original always had stability issues. I couldn’t  make it prime time ready. 
And now I maybe even can use it on pico-w‘s. My to do list just got bigger :)
Thanks for the hint!!!
